File: football-api/app/services/team_builder.py
import math
import random

BIB_COLOR_HEX: dict[str, str] = {
    "laranja": "#f97316",
    "azul": "#3b82f6",
    "verde": "#22c55e",
    "vermelho": "#ef4444",
    "amarelo": "#eab308",
    "preto": "#1f2937",
    "branco": "#f1f5f9",
}

TEAM_NAMES = [
    "Real Madruga", "Barcelusa", "Barsemlona", "Meia Boca Juniors",
    "Baile de Munique", "Varmeiras", "Atecubanos FC", "Inter de Limão",
    "Manchester Cachaça", "Real Matismo", "Paysanduba", "Horriver Plate",
    "Patético de Madrid", "Shakhtar dos Leks", "Espressinho da Mooca",
]

TEAM_COLORS = [
    "#e53e3e",  # vermelho
    "#3b82f6",  # azul
    "#f59e0b",  # amarelo
    "#22c55e",  # verde
    "#f97316",  # laranja
    "#a855f7",  # roxo
    "#ec4899",  # rosa
    "#06b6d4",  # ciano
    "#84cc16",  # limão
    "#14b8a6",  # verde-azulado
]


def _pick_names(n: int) -> list[str]:
    pool = TEAM_NAMES[:]
    random.shuffle(pool)
    names: list[str] = []
    used: set[str] = set()
    for name in pool:
        if name not in used:
            names.append(name)
            used.add(name)
        if len(names) == n:
            break
    idx = 2
    while len(names) < n:
        candidate = f"{pool[0]} {idx}"
        if candidate not in used:
            names.append(candidate)
            used.add(candidate)
            idx += 1
    return names


def _optimize_teams(times: list[list[dict]]) -> None:
    """
    Fase de otimização pós-distribuição: greedy swap para equalizar totais.

    Itera encontrando o melhor par de jogadores (um de cada time) cuja troca
    mais reduz a diferença entre o time com maior e menor skill_total.
    Para quando nenhuma troca melhora ou o spread fica ≤ 1.
    """
    for _ in range(500):
        totals = [sum(p["skill_stars"] for p in t) for t in times]
        current_spread = max(totals) - min(totals)
        if current_spread <= 1:
            break

        best_improvement = 0
        best_swap: tuple[int, int, int, int] | None = None

        n = len(times)
        gk_counts = [sum(1 for p in t if p.get("position") == "gk") for t in times]

        for a in range(n):
            for b in range(a + 1, n):
                for i, pa in enumerate(times[a]):
                    for j, pb in enumerate(times[b]):
                        if pa["skill_stars"] == pb["skill_stars"]:
                            continue

                        # GK constraint: não remover o único goleiro de um time
                        pa_is_gk = pa.get("position") == "gk"
                        pb_is_gk = pb.get("position") == "gk"
                        if pa_is_gk != pb_is_gk:
                            if pa_is_gk and gk_counts[a] <= 1:
                                continue  # time A perderia seu único GK
                            if pb_is_gk and gk_counts[b] <= 1:
                                continue  # time B perderia seu único GK

                        new_totals = totals[:]
                        new_totals[a] = totals[a] - pa["skill_stars"] + pb["skill_stars"]
                        new_totals[b] = totals[b] - pb["skill_stars"] + pa["skill_stars"]
                        new_spread = max(new_totals) - min(new_totals)
                        improvement = current_spread - new_spread
                        if improvement > best_improvement:
                            best_improvement = improvement
                            best_swap = (a, i, b, j)

        if best_swap is None:
            break

        a, i, b, j = best_swap
        times[a][i], times[b][j] = times[b][j], times[a][i]


def build_teams(
    confirmed: list[dict],
    players_per_team: int,
    team_slots: list[dict] | None = None,
) -> tuple[list[dict], list[dict]]:
    """
    Recebe lista de confirmados com player_id, skill_stars, position.
    Retorna (teams, reserves).

    players_per_team = jogadores de LINHA por time (exclui goleiro).
    Tamanho total de cada time = players_per_team + 1 (linha + goleiro).

    Algoritmo — distribuição por posição com sorteio em faixas de estrelas
    seguida de otimização greedy:
    1. Separa jogadores por posição.
    2. Goleiros: sorteia aleatoriamente qual time recebe cada um (1 por time).
    3. Para cada posição de linha (lat, zag, mei, ata):
       - Ordena por estrelas desc.
       - Distribui em rodadas de n_times jogadores por vez ("faixas").
       - Dentro de cada faixa embaralha aleatoriamente antes de atribuir
         um jogador a cada time — jogadores de nível similar vão para times
         diferentes, mas qual time recebe qual é sorteado.
    4. A soma de faixas por posição é limitada ao campo disponível no time
       (para evitar times maiores que team_size).
    5. Excedentes preenchem os slots restantes, distribuídos da mesma forma.
    6. Jogadores além da capacidade total viram reservas.
    7. Fase de otimização: greedy swap entre times para minimizar a diferença
       de skill_total entre o time mais forte e o mais fraco.
    """
    team_size = players_per_team + 1
    # ceil garante que todos os confirmados entram em algum time.
    # Com 39 jogadores e team_size=10 → 4 times (3 completos + 1 com 9).
    n_times = math.ceil(len(confirmed) / team_size)
    if n_times < 2:
        raise ValueError("Confirmados insuficientes para montar times.")

    # Embaralha para que jogadores com mesmas estrelas tenham ordem aleatória
    pool = confirmed[:]
    random.shuffle(pool)

    # Separa por posição e ordena por estrelas desc dentro de cada grupo
    by_pos: dict[str, list[dict]] = {}
    for p in pool:
        pos = p.get("position") or "mei"
        by_pos.setdefault(pos, []).append(p)
    for pos in by_pos:
        by_pos[pos].sort(key=lambda p: p["skill_stars"], reverse=True)

    gks = by_pos.pop("gk", [])
    times: list[list[dict]] = [[] for _ in range(n_times)]
    overflow: list[dict] = []

    def assign_tiers(group: list[dict], per_team: int) -> None:
        """
        Distribui `per_team` rodadas de `group` (ordenado por estrelas desc)
        entre os times. A cada rodada, pega os próximos n_times jogadores
        (mesma faixa de estrelas), embaralha e atribui um a cada time.
        Isso garante que jogadores de nível similar vão para times diferentes,
        com o sorteio decidindo qual time recebe qual.
        Excedentes vão para overflow.
        """
        to_dist = group[: per_team * n_times]
        overflow.extend(group[per_team * n_times :])

        for round_num in range(per_team):
            tier = to_dist[round_num * n_times : (round_num + 1) * n_times]
            shuffled = tier[:]
            random.shuffle(shuffled)
            for team_idx, player in enumerate(shuffled):
                times[team_idx].append(player)

    # Passo 1: Goleiros — 1 por time, sorteado aleatoriamente
    gks_for_teams = gks[:n_times]
    random.shuffle(gks_for_teams)
    for team_idx, gk in enumerate(gks_for_teams):
        times[team_idx].append(gk)
    overflow.extend(gks[n_times:])

    # Passo 2: Calcula per_team por posição, limitado ao campo disponível
    #
    # A soma de todos os per_team não pode ultrapassar field_slots, senão os
    # times ficam maiores que team_size. Reduz iterativamente a posição mais
    # abundante até caber.
    field_slots = team_size - 1
    positions = ["lat", "zag", "mei", "ata"]

    pos_per_team: dict[str, int] = {
        pos: len(by_pos.get(pos, [])) // n_times
        for pos in positions
    }

    while sum(pos_per_team.values()) > field_slots:
        max_pos = max(
            (p for p in positions if pos_per_team[p] > 0),
            key=lambda p: pos_per_team[p],
        )
        pos_per_team[max_pos] -= 1

    # Passo 3: Distribui cada posição por faixas embaralhadas
    for pos in positions:
        group = by_pos.get(pos, [])
        assign_tiers(group, pos_per_team[pos])

    # Passo 4: Overflow preenche slots restantes
    overflow.sort(key=lambda p: p["skill_stars"], reverse=True)
    remaining = [team_size - len(t) for t in times]
    final_reserves: list[dict] = []

    # 4a: Goleiros excedentes só vão para times que ainda não têm GK;
    #     caso contrário viram reservas (evita 2 goleiros no mesmo time)
    overflow_gks = [p for p in overflow if p.get("position") == "gk"]
    overflow_field = [p for p in overflow if p.get("position") != "gk"]

    random.shuffle(overflow_gks)
    for gk in overflow_gks:
        ti = next(
            (
                i for i, t in enumerate(times)
                if remaining[i] > 0 and not any(p.get("position") == "gk" for p in t)
            ),
            None,
        )
        if ti is not None:
            times[ti].append(gk)
            remaining[ti] -= 1
        else:
            final_reserves.append(gk)

    # 4b: Jogadores de linha preenchem os slots restantes por faixas embaralhadas.
    # Embaralha antes de ordenar para que jogadores com mesmas estrelas tenham
    # ordem aleatória entre grupos de posição — evita que os mesmos jogadores
    # fracos sejam sempre reservas por chegarem sempre no fim do array.
    overflow_field_sorted = overflow_field[:]
    random.shuffle(overflow_field_sorted)
    overflow_field_sorted.sort(key=lambda p: p["skill_stars"], reverse=True)

    idx = 0
    while idx < len(overflow_field_sorted):
        open_teams = [i for i in range(n_times) if remaining[i] > 0]
        if not open_teams:
            break
        batch = overflow_field_sorted[idx : idx + len(open_teams)]
        shuffled = batch[:]
        random.shuffle(shuffled)
        for ti, player in zip(open_teams, shuffled):
            times[ti].append(player)
            remaining[ti] -= 1
        idx += len(batch)

    final_reserves.extend(overflow_field_sorted[idx:])
    reserves = final_reserves

    # Fase de otimização: equaliza skill_total entre os times via greedy swap
    _optimize_teams(times)

    random_names = _pick_names(n_times)
    default_colors = TEAM_COLORS * ((n_times // len(TEAM_COLORS)) + 1)

    # Determine if any slot is configured (at least one non-empty slot)
    has_slots = bool(team_slots)

    result = []
    for pos_idx, players in enumerate(times, start=1):
        i = pos_idx - 1
        slot = team_slots[i] if has_slots and i < len(team_slots) else None

        if slot is not None:
            slot_name = slot.get("name") if isinstance(slot, dict) else getattr(slot, "name", None)
            slot_color_slug = slot.get("color") if isinstance(slot, dict) else getattr(slot, "color", None)
        else:
            slot_name = None
            slot_color_slug = None

        # Name priority: custom name > "Time {Cor}" > random
        if slot_name:
            name = slot_name
        elif slot_color_slug:
            name = f"Time {slot_color_slug.capitalize()}"
        else:
            name = random_names[i]

        # Color priority: slot hex > default
        if slot_color_slug and slot_color_slug in BIB_COLOR_HEX:
            color = BIB_COLOR_HEX[slot_color_slug]
        else:
            color = default_colors[i]

        result.append(
            {
                "name": name,
                "color": color,
                "position": pos_idx,
                "players": players,
                "skill_total": sum(p["skill_stars"] for p in players),
            }
        )

    return result, reserves

File: football-api-go/internal/services/team_builder.go
package services

import (
	"errors"
	"math"
	"math/rand"

	"github.com/thiagotn/football-manager/football-api-go/internal/db"
)

var bibColorHex = map[string]string{
	"laranja":  "#f97316",
	"azul":     "#3b82f6",
	"verde":    "#22c55e",
	"vermelho": "#ef4444",
	"amarelo":  "#eab308",
	"preto":    "#1f2937",
	"branco":   "#f1f5f9",
}

var teamNames = []string{
	"Real Madruga", "Barcelusa", "Barsemlona", "Meia Boca Juniors",
	"Baile de Munique", "Varmeiras", "Atecubanos FC", "Inter de Limão",
	"Manchester Cachaça", "Real Matismo", "Paysanduba", "Horriver Plate",
	"Patético de Madrid", "Shakhtar dos Leks", "Espressinho da Mooca",
}

var teamColors = []string{
	"#e53e3e", "#3b82f6", "#f59e0b", "#22c55e", "#f97316",
	"#a855f7", "#ec4899", "#06b6d4", "#84cc16", "#14b8a6",
}

// TeamResult holds the output of the team builder for one team.
type TeamResult struct {
	Name       string             `json:"name"`
	Color      string             `json:"color"`
	Position   int                `json:"position"`
	Players    []db.PlayerForDraw `json:"players"`
	SkillTotal int                `json:"skill_total"`
}

func pickNames(n int) []string {
	pool := make([]string, len(teamNames))
	copy(pool, teamNames)
	rand.Shuffle(len(pool), func(i, j int) { pool[i], pool[j] = pool[j], pool[i] })

	names := make([]string, 0, n)
	seen := make(map[string]bool)
	for _, name := range pool {
		if !seen[name] {
			names = append(names, name)
			seen[name] = true
		}
		if len(names) == n {
			break
		}
	}
	for idx := 2; len(names) < n; idx++ {
		candidate := pool[0] + " " + string(rune('0'+idx))
		if !seen[candidate] {
			names = append(names, candidate)
			seen[candidate] = true
		}
	}
	return names
}

func optimizeTeams(teams [][]db.PlayerForDraw) {
	for iter := 0; iter < 500; iter++ {
		totals := make([]int, len(teams))
		for i, t := range teams {
			for _, p := range t {
				totals[i] += p.SkillStars
			}
		}
		maxT, minT := totals[0], totals[0]
		for _, v := range totals {
			if v > maxT {
				maxT = v
			}
			if v < minT {
				minT = v
			}
		}
		if maxT-minT <= 1 {
			break
		}

		gkCounts := make([]int, len(teams))
		for i, t := range teams {
			for _, p := range t {
				if p.Position == "gk" {
					gkCounts[i]++
				}
			}
		}

		bestImprovement := 0
		bestA, bestI, bestB, bestJ := -1, -1, -1, -1
		currentSpread := maxT - minT

		for a := 0; a < len(teams); a++ {
			for b := a + 1; b < len(teams); b++ {
				for i, pa := range teams[a] {
					for j, pb := range teams[b] {
						if pa.SkillStars == pb.SkillStars {
							continue
						}
						paGK := pa.Position == "gk"
						pbGK := pb.Position == "gk"
						if paGK != pbGK {
							if paGK && gkCounts[a] <= 1 {
								continue
							}
							if pbGK && gkCounts[b] <= 1 {
								continue
							}
						}
						newA := totals[a] - pa.SkillStars + pb.SkillStars
						newB := totals[b] - pb.SkillStars + pa.SkillStars
						newMax, newMin := newA, newA
						for k, v := range totals {
							var vv int
							switch k {
							case a:
								vv = newA
							case b:
								vv = newB
							default:
								vv = v
							}
							if vv > newMax {
								newMax = vv
							}
							if vv < newMin {
								newMin = vv
							}
						}
						improvement := currentSpread - (newMax - newMin)
						if improvement > bestImprovement {
							bestImprovement = improvement
							bestA, bestI, bestB, bestJ = a, i, b, j
						}
					}
				}
			}
		}
		if bestA == -1 {
			break
		}
		teams[bestA][bestI], teams[bestB][bestJ] = teams[bestB][bestJ], teams[bestA][bestI]
	}
}

// BuildTeams implements the snake-draft + greedy-swap algorithm ported from Python.
// confirmed: slice of confirmed players with position and skill_stars.
// playersPerTeam: field players per team (GK not counted).
// teamSlots: optional custom name/color configuration from the group.
func BuildTeams(confirmed []db.PlayerForDraw, playersPerTeam int, teamSlots []db.TeamSlot) ([]TeamResult, []db.PlayerForDraw, error) {
	teamSize := playersPerTeam + 1
	nTeams := int(math.Ceil(float64(len(confirmed)) / float64(teamSize)))
	if nTeams < 2 {
		return nil, nil, errors.New("confirmed insuficientes para montar times")
	}

	// Shuffle then group by position
	pool := make([]db.PlayerForDraw, len(confirmed))
	copy(pool, confirmed)
	rand.Shuffle(len(pool), func(i, j int) { pool[i], pool[j] = pool[j], pool[i] })

	byPos := make(map[string][]db.PlayerForDraw)
	for _, p := range pool {
		pos := p.Position
		if pos == "" {
			pos = "mei"
		}
		byPos[pos] = append(byPos[pos], p)
	}
	for pos := range byPos {
		ps := byPos[pos]
		rand.Shuffle(len(ps), func(i, j int) { ps[i], ps[j] = ps[j], ps[i] })
		// sort by skill desc (stable within same skill due to shuffle above)
		for i := 0; i < len(ps)-1; i++ {
			for j := i + 1; j < len(ps); j++ {
				if ps[j].SkillStars > ps[i].SkillStars {
					ps[i], ps[j] = ps[j], ps[i]
				}
			}
		}
		byPos[pos] = ps
	}

	gks := byPos["gk"]
	delete(byPos, "gk")

	teams := make([][]db.PlayerForDraw, nTeams)
	for i := range teams {
		teams[i] = []db.PlayerForDraw{}
	}
	var overflow []db.PlayerForDraw

	assignTiers := func(group []db.PlayerForDraw, perTeam int) {
		toDist := group
		if len(toDist) > perTeam*nTeams {
			overflow = append(overflow, toDist[perTeam*nTeams:]...)
			toDist = toDist[:perTeam*nTeams]
		}
		for round := 0; round < perTeam; round++ {
			start := round * nTeams
			end := start + nTeams
			if end > len(toDist) {
				end = len(toDist)
			}
			tier := make([]db.PlayerForDraw, end-start)
			copy(tier, toDist[start:end])
			rand.Shuffle(len(tier), func(i, j int) { tier[i], tier[j] = tier[j], tier[i] })
			for ti, p := range tier {
				teams[ti] = append(teams[ti], p)
			}
		}
	}

	// Step 1: assign GKs
	gksForTeams := gks
	if len(gksForTeams) > nTeams {
		overflow = append(overflow, gksForTeams[nTeams:]...)
		gksForTeams = gksForTeams[:nTeams]
	}
	rand.Shuffle(len(gksForTeams), func(i, j int) { gksForTeams[i], gksForTeams[j] = gksForTeams[j], gksForTeams[i] })
	for ti, gk := range gksForTeams {
		teams[ti] = append(teams[ti], gk)
	}

	// Step 2: per-position allocation
	fieldSlots := teamSize - 1
	positions := []string{"lat", "zag", "mei", "ata"}
	posPerTeam := make(map[string]int, len(positions))
	for _, pos := range positions {
		if g := byPos[pos]; len(g) > 0 {
			posPerTeam[pos] = len(g) / nTeams
		}
	}
	for {
		total := 0
		for _, v := range posPerTeam {
			total += v
		}
		if total <= fieldSlots {
			break
		}
		maxPos := ""
		maxVal := 0
		for _, pos := range positions {
			if posPerTeam[pos] > maxVal {
				maxVal = posPerTeam[pos]
				maxPos = pos
			}
		}
		posPerTeam[maxPos]--
	}

	for _, pos := range positions {
		if g := byPos[pos]; len(g) > 0 {
			assignTiers(g, posPerTeam[pos])
		}
	}

	// Step 3: overflow fills remaining slots
	remaining := make([]int, nTeams)
	for i, t := range teams {
		remaining[i] = teamSize - len(t)
	}

	var finalReserves []db.PlayerForDraw

	// GK overflow: only to teams without a GK
	var ovGKs, ovField []db.PlayerForDraw
	for _, p := range overflow {
		if p.Position == "gk" {
			ovGKs = append(ovGKs, p)
		} else {
			ovField = append(ovField, p)
		}
	}
	rand.Shuffle(len(ovGKs), func(i, j int) { ovGKs[i], ovGKs[j] = ovGKs[j], ovGKs[i] })
	for _, gk := range ovGKs {
		assigned := false
		for i, t := range teams {
			if remaining[i] > 0 {
				hasGK := false
				for _, p := range t {
					if p.Position == "gk" {
						hasGK = true
						break
					}
				}
				if !hasGK {
					teams[i] = append(teams[i], gk)
					remaining[i]--
					assigned = true
					break
				}
			}
		}
		if !assigned {
			finalReserves = append(finalReserves, gk)
		}
	}

	// Field overflow
	rand.Shuffle(len(ovField), func(i, j int) { ovField[i], ovField[j] = ovField[j], ovField[i] })
	// sort by skill desc (stable: shuffle first)
	for i := 0; i < len(ovField)-1; i++ {
		for j := i + 1; j < len(ovField); j++ {
			if ovField[j].SkillStars > ovField[i].SkillStars {
				ovField[i], ovField[j] = ovField[j], ovField[i]
			}
		}
	}
	idx := 0
	for idx < len(ovField) {
		var openTeams []int
		for i, r := range remaining {
			if r > 0 {
				openTeams = append(openTeams, i)
			}
		}
		if len(openTeams) == 0 {
			break
		}
		end := idx + len(openTeams)
		if end > len(ovField) {
			end = len(ovField)
		}
		batch := make([]db.PlayerForDraw, end-idx)
		copy(batch, ovField[idx:end])
		rand.Shuffle(len(batch), func(i, j int) { batch[i], batch[j] = batch[j], batch[i] })
		for k, p := range batch {
			if k < len(openTeams) {
				ti := openTeams[k]
				teams[ti] = append(teams[ti], p)
				remaining[ti]--
			}
		}
		idx += len(batch)
	}
	finalReserves = append(finalReserves, ovField[idx:]...)

	// Optimization: greedy swap
	optimizeTeams(teams)

	randomNames := pickNames(nTeams)
	defaultColors := make([]string, nTeams)
	for i := range defaultColors {
		defaultColors[i] = teamColors[i%len(teamColors)]
	}

	result := make([]TeamResult, nTeams)
	for i, players := range teams {
		pos := i + 1
		var slot *db.TeamSlot
		if teamSlots != nil && i < len(teamSlots) {
			slot = &teamSlots[i]
		}

		var name, color string
		if slot != nil && slot.Name != nil && *slot.Name != "" {
			name = *slot.Name
		} else if slot != nil && slot.Color != nil && *slot.Color != "" {
			// Capitalize first letter
			s := *slot.Color
			if len(s) > 0 {
				name = "Time " + string(s[0]-32) + s[1:]
			} else {
				name = randomNames[i]
			}
		} else {
			name = randomNames[i]
		}

		if slot != nil && slot.Color != nil {
			if hex, ok := bibColorHex[*slot.Color]; ok {
				color = hex
			} else {
				color = defaultColors[i]
			}
		} else {
			color = defaultColors[i]
		}

		skillTotal := 0
		for _, p := range players {
			skillTotal += p.SkillStars
		}

		result[i] = TeamResult{
			Name:       name,
			Color:      color,
			Position:   pos,
			Players:    players,
			SkillTotal: skillTotal,
		}
	}

	return result, finalReserves, nil
}
